core_contrib/md5.shs/shs.1
.\" @(#)shs.1	12.4 9/1/96 11:26:26
.\"
.\" This man page was written by:
.\"
.\"	 Landon Curt Noll  (chongo@toad.com)	chongo <was here> /\../\
.\"
.\" This man page has been placed in the public domain.  Please do not 
.\" copyright this man page.
.\"
.\" LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH  REGARD  TO
.\" THIS  SOFTWARE,  INCLUDING  ALL IMPLIED WARRANTIES OF MER-
.\" CHANTABILITY AND FITNESS.  IN NO EVENT SHALL  LANDON  CURT
.\" NOLL  BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
.\" DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM  LOSS  OF
.\" USE,  DATA  OR  PROFITS, WHETHER IN AN ACTION OF CONTRACT,
.\" NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR  IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" See shsdrvr.c for version and modification history.
.\"
.TH SHS 1  LOCAL
.SH NAME
shs1, shs, sha1, sha \- Secure Hash Standard
.SH SYNOPSIS
.B shs1
[\fB\-cCdhiqtx\fP]
[\fB\-m num\fP]
.br
     [\fB\-p prefix\fP]
[\fB\-P prefile\fP]
[\fB\-s str\fP]
file\ ...
.sp
.B shs
[\fB\-cCdhiqtx\fP]
[\fB\-m num\fP]
.br
    [\fB\-p prefix\fP]
[\fB\-P prefile\fP]
[\fB\-s str\fP]
file\ ...
.sp
.B sha1
[\fB\-cCdhiqtx\fP]
[\fB\-m num\fP]
.br
     [\fB\-p prefix\fP]
[\fB\-P prefile\fP]
[\fB\-s str\fP]
file\ ...
.sp
.B sha
[\fB\-cCdhiqtx\fP]
[\fB\-m num\fP]
.br
    [\fB\-p prefix\fP]
[\fB\-P prefile\fP]
[\fB\-s str\fP]
file\ ...
.SH DESCRIPTION
The
.B sha1
utility implements the Secure Hash Algorithm-1 (\fISHA-1\fP).
It produces 160-bit Secure Hash Digests of files, strings or data 
read on stdin.
If no
.B file
is given, and
no string is to be digested (\fB\-s\fP)
then stdin will be digested.
By default, a digests are printed as 40 hex characters without a leading 
.IR 0x .
.PP
The
.B shs1
utility is the same as
.B sha1
and is linked for backward compatibility.
.PP
The Secure Hash Standard-1 (\fISHS-1\fP) is a United States
Department of Commerce National Institute of Standards and Technology
approved standard (FIPS Pub 180-1) for secure hashing.
The 
.I SHS-1
is designed to work with the Digital Signature Standard (\fIDSS\fP).
The Secure Hash Algorithm (\fISHA-1\fP) 
specified in this standard is
necessary to ensure the security of the Digital Signature Standard. 
When a message of length < 2^64 bits is input, the 
.I SHA-1
produces a 160-bit representation of the message called the message digest. 
The message digest is used during generation of a signature for the message.  
The
.I SHA-1
is designed to have the following properties: it is
computationally infeasible to recover a message corresponding to a
given message digest, or to find two different messages which
produce the same message digest.
.PP
On 1994 May 31, the United States Department of Commerce 
National Institute of Standards and Technology published a
technical modification (FIPS Pub 180-1) to the Secure Hash Standard known
as Secure Hash Algorithm-1 (\fISHA-1\fP).
The
.B sha1
utility implements the new Secure Hash Algorithm-1 (\fISHA-1\fP)
as specified by (FIPS Pub 180-1).
The
.B sha
utility implements the old Secure Hash Algorithm (\fISHA\fP)
as specified by (FIPS Pub 180).
It uses the same arguments and interface as
.BR sha1 .
This utility is provided for backward compatibility with
version at or before 2.10.1.
.PP
The
.B shs
utility is the same as
.B sha
and is linked for backward compatibility.
.PP
If a 
.B str 
(string) argument is given, then the digest for 
.BR str ,
followed by a space, followed by
.B str 
enclosed double quotes is written to stdout.
Any
.B file
arguments are ignored.
.PP
If one or more
.B file
if given, a separate digest if produced for each file.
By default, 
.B file 
digests are followed by a space and the filename.
.PP
If no 
.B str
or
.B file
arguments are given, then a digest of stdin is written to stdout.
.PP
.TP
.B \-c
Print C style hex digests with a leading \fI0x\fP.
.TP
.B \-C
When printing multi digests (see \fB\-d\fP or \fB\-m num\fP), do not
separate digests with spaces.  When combined with \fB\-c\fP and \fB\-q\fP,
the output is only a single long hex value.
.TP
.B \-d
Shorthand for \fB\-m 2\fP.
Used for backward compatibility.
.TP
.B \-h
Print a help and usage message.
.TP
.B \-i
Compute inode digests of files.
.sp
Prepend the filename and various inode information to the file data 
being digested.
The inode information prepended includes information
such as the device (\fIst_dev\fP), 
inode number (\fIst_ino\fP), 
mode (\fIst_mode\fP), 
link count (\fIst_nlink\fP), 
uid (\fIst_uid\fP), 
gid (\fIst_gid\fP), 
size (\fIst_size\fP),
modification time (\fIst_mtime\fP) and
change time (\fIst_ctime\fP).
The prepended data is padded with zeros to make it a multiple of 64
bytes long.
Both a
.BR stat (2)
and a
.BR lstat (2)
information are both used.
.sp
The \fB\-i\fP flag allows one to include various inode information in
the digest.  
This option is useful in detecting file tampering.
For example, the following will produce different digests:
.sp
.in +0.5i
.nf
sha1 \-i /tmp/chongo
cp /tmp/chongo /tmp/was_here
\^... misc funny business ...
cp /tmp/was_here /tmp/chongo
sha1 \-i /tmp/chongo
.fi
.in -0.5i
.sp
Note that a \fB\-i\fP digest is not portable to other systems.
This is because inode information will likely change as the
contents of a file are copied from machine to machine.
.sp
To mark a \fB\-i\fP digest as a special value, \fI.0\fP is 
prepended onto the digest output.
.sp
One may only compute inode digests of files.
Use of
.B \-i
disables reading from stdin.
The
.B \-i
flag is not compatible with
.BR \-s .
.TP
.BI \-m num
Compute multiple digests in parallel.
.sp
Every \fBnum\fP-th octet of the prefix (if given) and data are processed
by a separate digest.
Each successive octet is assigned an index starting with 0.
The i-th digest is taken from all octets with an index of i mod \fBnum\fP.
.sp
The result of a multiple digest is a hash that is \fBnum\fP times
as long as a standard digest.
.TP
.BI \-p prefix
.TP
.BI \-P prefile
Insert a prefix into the data to be digested.
.sp
By using \fB\-p\fP\fIprefix\fP, one may prepend any set of data with a
string.  
The digest produced is equivalent to digest that is
produced with the string pretended to the data.  
Thus the following two commands produce the same digest:
.sp
.in +0.5i
.nf
sha1 \-p curds \-s whey
sha1 \-s curdswhey
.fi
.in -0.5i
.sp
By use of the \fB\-P\fP\fIprefile\fP interface, one may prepend using
up to the 32k of a file.  
This interface allows one to prepend using binary data.  
The following produces the same digest:
.sp
.in +0.5i
.nf
sha1 \-P /usr/bin/awk /bin/ls
dd if=/usr/bin/awk of=/tmp/foo bs=32k count=1
cat /tmp/foo /bin/ls > /tmp/ls
sha1 /tmp/ls
.fi
.in -0.5i
.sp
Knowledge of the original prepend data is only kept in
the digest.  
Thus one may use the prepend string as a ``salt'' making it intractable 
for someone else to reproduce the digest of a file without knowledge 
of the prepend data.
.sp
This feature works in conjunction all modes of operation except
the \fB\-t\fP and \fB\-x\fP modes.
.TP
.B \-q
Output only digests.
The filename or string will not be written to stdout.
.TP
.BR \-s str
Digest \fIstr\fP as if it were a string.
.sp
The trailing NUL byte is not digested.
No files are digested.
The \fIstr\fP is written to stdout enclosed in double quotes.
.TP
.BR \-t
Time the user cpu seconds needed to digest several megabytes of data.
The number of megabytes,
followed by the digest produced,
followed by the number of user cpu seconds
followed by the number of characters per user second is written to stdout.
.sp
On some systems with certain virtual memory characteristics,
you may need to run the performance test several times, ignoring the
first result.
The default amount of data tested is 16 megabytes.
On some systems, this amount may differ.
One may change the number of megabytes processed by changing the value of
TEST_MEG in the file shs1drvr.c (or shsdrvr.c) and recompiling.
.TP
.BR \-v
Print the version.
.TP
.BR \-x
Perform an extended standard SHS-1 test suite.
.sp
The test suite will first look in the current directory for
.IR file1 .
If it is not found, it will look in ${DESTDIR} (usually 
.IR /usr/local/lib/shs ).
It is assumes that 
.I file2 
is in the same location as 
.I file1.
.sp
The standard test suite been extended.  
The initial string now says:
.sp
.in +0.5i
.nf
sha1 test suite results
.fi
.in -0.5i
.sp
instead of:
.sp
.in +0.5i
.nf
SHA test suite results
.fi
.in -0.5i
.sp
to reflect the utility name instead of the algorithm implemented.
This also helps distinguish this version from much older 
versions which did not have the 
.B -v 
flag.
Also the original test file
.I foo
which contained the string  "\fIabc\fP" (with no newline) was renamed 
.IR file1 .
.SH "SEE ALSO"
.BR md5 (1),
.BR stat (2),
.BR lstat (2)
.SH FILES
.nf
\^./file1			default test file location
\^./file2			default test file location
${DESTDIR}/file1	alternate test file location
${DESTDIR}/file2	alternate test file location
.fi
.sp
The typical value of ${DESTDIR} is \fI/usr/local/lib/shs\fP,
or \fI/usr/local/lib\fP.
.SH AUTHOR
.nf
Much of this code was written, re-written or modified by:

    Landon Curt Noll	(chongo@toad.com)  /\\../\\

This code is based on code by Peter C. Gutmann.  Much thanks goes 
to Peter C. Gutman (pgut1@cs.aukuni.ac.nz) , Shawn A. Clifford
(sac@eng.ufl.edu), Pat Myrto (pat@rwing.uucp), Colin Plumb
(colin@nyx10.cs.du.edu), Rich Schroeppel (rcs@cs.arizona.edu)
and others who wrote and/or worked on the original code.
.fi
.SH NOTICE
LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. 
IN NO EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
.SH BUGS
The command:
.sp
.in +0.5i
.nf
sha1 -s '/\\"O/\\'
.fi
.in -0.5i
.sp
writes to stdout, the following line:
.sp
.in +0.5i
.nf
945a6306a6d7caee4a28fabb36838750a673fc9f "/\"O/\"
.fi
.in -0.5i
.sp
The string written in double quotes by
.B \-s
may not be a valid string according to C or shell syntax.


core_contrib/md5.shs/shs.c
/* @(#)shs.c	12.1 17 Nov 1995 04:22:20 */
/*
 * shs - old Secure Hash Standard
 *
 **************************************************************************
 * This version implements the old Secure Hash Algorithm specified by     *
 * (FIPS Pub 180).  This version is kept for backward compatibility with  *
 * shs version 2.10.1.  See the shs utility for the new standard.	  *
 **************************************************************************
 *
 * Written 2 September 1992, Peter C. Gutmann.
 *
 * This file was Modified/Re-written by:
 *
 *	 Landon Curt Noll  (chongo@toad.com)	chongo <was here> /\../\
 *
 * This code has been placed in the public domain.  Please do not
 * copyright this code.
 *
 * LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH  REGARD  TO
 * THIS  SOFTWARE,  INCLUDING  ALL IMPLIED WARRANTIES OF MER-
 * CHANTABILITY AND FITNESS.  IN NO EVENT SHALL  LANDON  CURT
 * NOLL  BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM  LOSS  OF
 * USE,  DATA  OR  PROFITS, WHETHER IN AN ACTION OF CONTRACT,
 * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR  IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * See shsdrvr.c for version and modification history.
 */

#include <stdio.h>
#include <string.h>
#define SHS_IO
#include "shs.h"
#include "align.h"
#include "endian.h"

char *shs_what="@(#)";	/* #(@) if checked in */

/*
 * The SHS f()-functions.  The f1 and f3 functions can be optimized
 * to save one boolean operation each - thanks to Rich Schroeppel,
 * rcs@cs.arizona.edu for discovering this.
 *
 * f1: ((x&y) | (~x&z)) == (z ^ (x&(y^z)))
 * f3: ((x&y) | (x&z) | (y&z)) == ((x&y) | (z&(x|y)))
 */
#define f1(x,y,z)       (z ^ (x&(y^z)))		/* Rounds  0-19 */
#define f2(x,y,z)       (x^y^z)			/* Rounds 20-39 */
#define f3(x,y,z)       ((x&y) | (z&(x|y)))	/* Rounds 40-59 */
#define f4(x,y,z)       (x^y^z)			/* Rounds 60-79 */

/* The SHS Mysterious Constants */
#define K1      0x5A827999L	/* Rounds  0-19 */
#define K2      0x6ED9EBA1L	/* Rounds 20-39 */
#define K3      0x8F1BBCDCL	/* Rounds 40-59 */
#define K4      0xCA62C1D6L	/* Rounds 60-79 */

/* SHS initial values */
#define h0init  0x67452301L
#define h1init  0xEFCDAB89L
#define h2init  0x98BADCFEL
#define h3init  0x10325476L
#define h4init  0xC3D2E1F0L

/* 32-bit rotate left - kludged with shifts */
#define LEFT_ROT(X,n)  (((X)<<(n)) | ((X)>>(32-(n))))

/*
 * The initial expanding function.  The hash function is defined over an
 * 80-word expanded input array W, where the first 16 are copies of the input
 * data, and the remaining 64 are defined by
 *
 *      W[i] = W[i-16] ^ W[i-14] ^ W[i-8] ^ W[i-3]
 *
 * This implementation generates these values on the fly in a circular
 * buffer - thanks to Colin Plumb (colin@nyx10.cs.du.edu) for this
 * optimization.
 */
#define exor(W,i) (W[i&15] ^= (W[(i-14)&15] ^ W[(i-8)&15] ^ W[(i-3)&15]))

/*
 * The prototype SHS sub-round.  The fundamental sub-round is:
 *
 *      a' = e + LEFT_ROT(a,5) + f(b,c,d) + k + data;
 *      b' = a;
 *      c' = LEFT_ROT(b,30);
 *      d' = c;
 *      e' = d;
 *
 * but this is implemented by unrolling the loop 5 times and renaming the
 * variables ( e, a, b, c, d ) = ( a', b', c', d', e' ) each iteration.
 * This code is then replicated 20 times for each of the 4 functions, using
 * the next 20 values from the W[] array each time.
 */
#define subRound(a, b, c, d, e, f, k, data) \
    (e += LEFT_ROT(a,5) + f(b,c,d) + k + data, b = LEFT_ROT(b,30))

/* forward declarations */
static void shsTransform P((ULONG*, ULONG*));
static ULONG in[SHS_CHUNKWORDS];


/*
 * shsInit - initialize the SHS state
 */
void
shsInit(dig)
    SHS_INFO *dig;
{
    /* Set the h-vars to their initial values */
    dig->digest[0] = h0init;
    dig->digest[1] = h1init;
    dig->digest[2] = h2init;
    dig->digest[3] = h3init;
    dig->digest[4] = h4init;

    /* Initialise bit count */
    dig->countLo = 0;
    dig->countHi = 0;
    dig->datalen = 0;
}


/*
 * shsTransform - perform the SHS transformatio
 *
 * Note that this code, like MD5, seems to break some optimizing compilers.
 * It may be necessary to split it into sections, eg based on the four
 * subrounds.  One may also want to roll each subround into a loop.
 */
static void
shsTransform(digest, W)
    ULONG *digest;
    ULONG *W;
{
    ULONG A, B, C, D, E;	/* Local vars */

    /* Set up first buffer and local data buffer */
    A = digest[0];
    B = digest[1];
    C = digest[2];
    D = digest[3];
    E = digest[4];

    /* Heavy mangling, in 4 sub-rounds of 20 interations each. */
    subRound(A, B, C, D, E, f1, K1, W[ 0]);
    subRound(E, A, B, C, D, f1, K1, W[ 1]);
    subRound(D, E, A, B, C, f1, K1, W[ 2]);
    subRound(C, D, E, A, B, f1, K1, W[ 3]);
    subRound(B, C, D, E, A, f1, K1, W[ 4]);
    subRound(A, B, C, D, E, f1, K1, W[ 5]);
    subRound(E, A, B, C, D, f1, K1, W[ 6]);
    subRound(D, E, A, B, C, f1, K1, W[ 7]);
    subRound(C, D, E, A, B, f1, K1, W[ 8]);
    subRound(B, C, D, E, A, f1, K1, W[ 9]);
    subRound(A, B, C, D, E, f1, K1, W[10]);
    subRound(E, A, B, C, D, f1, K1, W[11]);
    subRound(D, E, A, B, C, f1, K1, W[12]);
    subRound(C, D, E, A, B, f1, K1, W[13]);
    subRound(B, C, D, E, A, f1, K1, W[14]);
    subRound(A, B, C, D, E, f1, K1, W[15]);
    subRound(E, A, B, C, D, f1, K1, exor(W,16));
    subRound(D, E, A, B, C, f1, K1, exor(W,17));
    subRound(C, D, E, A, B, f1, K1, exor(W,18));
    subRound(B, C, D, E, A, f1, K1, exor(W,19));

    subRound(A, B, C, D, E, f2, K2, exor(W,20));
    subRound(E, A, B, C, D, f2, K2, exor(W,21));
    subRound(D, E, A, B, C, f2, K2, exor(W,22));
    subRound(C, D, E, A, B, f2, K2, exor(W,23));
    subRound(B, C, D, E, A, f2, K2, exor(W,24));
    subRound(A, B, C, D, E, f2, K2, exor(W,25));
    subRound(E, A, B, C, D, f2, K2, exor(W,26));
    subRound(D, E, A, B, C, f2, K2, exor(W,27));
    subRound(C, D, E, A, B, f2, K2, exor(W,28));
    subRound(B, C, D, E, A, f2, K2, exor(W,29));
    subRound(A, B, C, D, E, f2, K2, exor(W,30));
    subRound(E, A, B, C, D, f2, K2, exor(W,31));
    subRound(D, E, A, B, C, f2, K2, exor(W,32));
    subRound(C, D, E, A, B, f2, K2, exor(W,33));
    subRound(B, C, D, E, A, f2, K2, exor(W,34));
    subRound(A, B, C, D, E, f2, K2, exor(W,35));
    subRound(E, A, B, C, D, f2, K2, exor(W,36));
    subRound(D, E, A, B, C, f2, K2, exor(W,37));
    subRound(C, D, E, A, B, f2, K2, exor(W,38));
    subRound(B, C, D, E, A, f2, K2, exor(W,39));

    subRound(A, B, C, D, E, f3, K3, exor(W,40));
    subRound(E, A, B, C, D, f3, K3, exor(W,41));
    subRound(D, E, A, B, C, f3, K3, exor(W,42));
    subRound(C, D, E, A, B, f3, K3, exor(W,43));
    subRound(B, C, D, E, A, f3, K3, exor(W,44));
    subRound(A, B, C, D, E, f3, K3, exor(W,45));
    subRound(E, A, B, C, D, f3, K3, exor(W,46));
    subRound(D, E, A, B, C, f3, K3, exor(W,47));
    subRound(C, D, E, A, B, f3, K3, exor(W,48));
    subRound(B, C, D, E, A, f3, K3, exor(W,49));
    subRound(A, B, C, D, E, f3, K3, exor(W,50));
    subRound(E, A, B, C, D, f3, K3, exor(W,51));
    subRound(D, E, A, B, C, f3, K3, exor(W,52));
    subRound(C, D, E, A, B, f3, K3, exor(W,53));
    subRound(B, C, D, E, A, f3, K3, exor(W,54));
    subRound(A, B, C, D, E, f3, K3, exor(W,55));
    subRound(E, A, B, C, D, f3, K3, exor(W,56));
    subRound(D, E, A, B, C, f3, K3, exor(W,57));
    subRound(C, D, E, A, B, f3, K3, exor(W,58));
    subRound(B, C, D, E, A, f3, K3, exor(W,59));

    subRound(A, B, C, D, E, f4, K4, exor(W,60));
    subRound(E, A, B, C, D, f4, K4, exor(W,61));
    subRound(D, E, A, B, C, f4, K4, exor(W,62));
    subRound(C, D, E, A, B, f4, K4, exor(W,63));
    subRound(B, C, D, E, A, f4, K4, exor(W,64));
    subRound(A, B, C, D, E, f4, K4, exor(W,65));
    subRound(E, A, B, C, D, f4, K4, exor(W,66));
    subRound(D, E, A, B, C, f4, K4, exor(W,67));
    subRound(C, D, E, A, B, f4, K4, exor(W,68));
    subRound(B, C, D, E, A, f4, K4, exor(W,69));
    subRound(A, B, C, D, E, f4, K4, exor(W,70));
    subRound(E, A, B, C, D, f4, K4, exor(W,71));
    subRound(D, E, A, B, C, f4, K4, exor(W,72));
    subRound(C, D, E, A, B, f4, K4, exor(W,73));
    subRound(B, C, D, E, A, f4, K4, exor(W,74));
    subRound(A, B, C, D, E, f4, K4, exor(W,75));
    subRound(E, A, B, C, D, f4, K4, exor(W,76));
    subRound(D, E, A, B, C, f4, K4, exor(W,77));
    subRound(C, D, E, A, B, f4, K4, exor(W,78));
    subRound(B, C, D, E, A, f4, K4, exor(W,79));

    /* Build message digest */
    digest[0] += A;
    digest[1] += B;
    digest[2] += C;
    digest[3] += D;
    digest[4] += E;
}


/*
 * shsUpdate - update SHS with arbitrary length data
 *
 * This code does not assume that the buffer size is a multiple of
 * SHS_CHUNKSIZE bytes long.  This code handles partial chunk between
 * calls to shsUpdate().
 */
void
shsUpdate(dig, buffer, count)
    SHS_INFO *dig;
    BYTE *buffer;
    ULONG count;
{
    ULONG datalen = dig->datalen;

    /*
     * Catch the case of a non-empty data buffer
     */
    if (datalen > 0) {

	/* determine the size we need to copy */
	ULONG cpylen = SHS_CHUNKSIZE - datalen;

	/* case: new data will not fill the buffer */
	if (cpylen > count) {
	    memcpy((char *)dig->data+datalen, (char *)buffer, count);
	    dig->datalen = datalen+count;
	    return;

	/* case: buffer will be filled */
	} else {
	    memcpy((char *)dig->data+datalen, (char *)buffer, cpylen);
	    SHS_TRANSFORM(dig, in, dig->data);
	    buffer += cpylen;
	    count -= cpylen;
	    dig->datalen = 0;
	}
    }

    /*
     * Process data in SHS_CHUNKSIZE chunks
     */
    if (count >= SHS_CHUNKSIZE) {
	shsfullUpdate(dig, buffer, count);
	buffer += (count/SHS_CHUNKSIZE)*SHS_CHUNKSIZE;
	count %= SHS_CHUNKSIZE;
    }

    /*
     * Handle any remaining bytes of data.
     * This should only happen once on the final lot of data
     */
    if (count > 0) {
	memcpy((char *)dig->data, (char *)buffer, count);
    }
    dig->datalen = count;
}


/*
 * shsfullUpdate - update SHS with chunk multiple length data
 *
 * This function assumes that count is a multiple of SHS_CHUNKSIZE and that
 * no partial chunk is left over from a previous call.
 */
void
shsfullUpdate(dig, buffer, count)
    SHS_INFO *dig;
    BYTE *buffer;
    ULONG count;
{
    /*
     * Process data in SHS_CHUNKSIZE chunks
     */
    while (count >= SHS_CHUNKSIZE) {
#if defined(MUST_ALIGN)
	if ((long)buffer & (sizeof(ULONG)-1)) {
	    memcpy((char *)in, (char *)buffer, SHS_CHUNKSIZE);
	    SHS_TRANSFORM(dig, in, in);
	} else {
	    SHS_TRANSFORM(dig, in, buffer);
	}
#else
	SHS_TRANSFORM(dig, in, buffer);
#endif
	buffer += SHS_CHUNKSIZE;
	count -= SHS_CHUNKSIZE;
    }
}


/*
 * shsFinal - perform final SHS transforms
 *
 * At this point we have less than a full chunk of data remaining
 * (and possibly no data) in the shs state data buffer.
 *
 * First we append a final 0x80 byte.
 *
 * Next if we have more than 56 bytes, we will zero fill the remainder
 * of the chunk, transform and then zero fill the first 56 bytes.
 * If we have 56 or fewer bytes, we will zero fill out to the 56th
 * chunk byte.  Regardless, we wind up with 56 bytes data.
 *
 * Finally we append the 64 bit length on to the 56 bytes of data
 * remaining.  This final chunk is transformed.
 */
void
shsFinal(dig)
    SHS_INFO *dig;
{
    int count = dig->datalen;
    ULONG lowBitcount = dig->countLo;
    ULONG highBitcount = dig->countHi;

    /*
     * Set the first char of padding to 0x80.
     * This is safe since there is always at least one byte free
     */
    ((BYTE *)dig->data)[count++] = 0x80;

    /* Pad out to 56 mod SHS_CHUNKSIZE */
    if (count > 56) {
	/* Two lots of padding:  Pad the first chunk to SHS_CHUNKSIZE bytes */
	memset((BYTE *)dig->data + count, 0, SHS_CHUNKSIZE - count);
	SHS_TRANSFORM(dig, dig->data, dig->data);

	/* Now fill the next chunk with 56 bytes */
	memset(dig->data, 0, 56);
    } else {
	/* Pad chunk to 56 bytes */
	memset((BYTE *)dig->data + count, 0, 56 - count);
    }
#if BYTE_ORDER == LITTLE_ENDIAN
    SHS_SWAP_BYTE_SEX(dig->data, dig->data);
#endif

    /*
     * Append length in bits and transform
     *
     * We assume that bit count is a multiple of 8 because we have
     * only processed full bytes.
     */
    dig->data[SHS_HIGH] = (highBitcount << 3) | (lowBitcount >> 29);
    dig->data[SHS_LOW] = (lowBitcount << 3);
    shsTransform(dig->digest, dig->data);
    dig->datalen = 0;
}
